### PR TITLE
fix updating nodes states

### PR DESCRIPTION
**Pull request by jeffnvidia**
## Summary
Based on main
This PR fixes a bug : cloudAI wasnt able to correctly assess the state of the nodes (IDLE, ALLOCATED ...)
fixed the parse_sinfo_output function

## Test Plan
1. Test by @amaslenn
CI
2. Test by @jeffnvidia
2.1 Slurm command generation for a given URL

$ cloudaix --mode run --system_config_path conf/v0.6/general/system/israel_1.toml --test_scenario_path conf/v0.6/general/test_scenario/llama/llama.toml

### CODE REVIEW

**Review by TaekyungHeo — CHANGES_REQUESTED** [review]
Could you please update
https://github.com/NVIDIA/cloudai/blob/main/tests/test_slurm_system.py
as well, so that we can understand which cases were not covered and are now covered?

File: tests/test_slurm_system.py [hunk 1]
```diff
@@ -48,6 +54,20 @@ def test_parse_sinfo_output(slurm_system):
...
+def test_parse_sinfo_output2(slurm_system):
+    sinfo_output = """
+    PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
+    main    up    3:00:00      1  inval node-081
+    main    up    3:00:00      5  drain node-[065-066,114,124-125]
+    main    up    3:00:00      2   resv node-[034-035]
+    main    up    3:00:00     88  alloc node-[033,036-064,067-080,082-113,115-123,126-128]
+    backup    up   12:00:00     16  alloc node-[01-16]
+    """
+    node_user_map = {'': 'user1', 'node-033': 'user2', 'node-[036-064': 'user3', '067-080': 'user3', '082-113': 'user3', '115-118]': 'user3', 'node-[119-123': 'user4', '126-128]': 'user4', 'node-01': 'user5', 'node-02': 'user5', 'node-03': 'user5', 'node-04': 'user5', 'node-05': 'user5', 'node-06': 'user5', 'node-07': 'user5', 'node-08': 'user5', 'node-09': 'user5', 'node-10': 'user5', 'node-11': 'user5', 'node-12': 'user5', 'node-13': 'user5', 'node-14': 'user5', 'node-15': 'user5', 'node-16': 'user5'}
```
Comment (TaekyungHeo) [inline tests/test_slurm_system.py:67]: I introduced `node_user_map` to create a mapping between nodes and users. A node has an assigned username (UNIX name) if the node is currently being used by a user. My concern about this test case is that we have a malformed node name, such as `node-[036-064`, without a closing bracket. Is this the case during execution? If so, we need to fix it as well.

File: tests/test_slurm_system.py [hunk 2]
```diff
@@ -8,8 +8,14 @@
 @pytest.fixture
 def slurm_system():
     nodes = [
-        SlurmNode(name="nodeA001", partition="main", state=SlurmNodeState.UNKNOWN_STATE),
-        SlurmNode(name="nodeB001", partition="main", state=SlurmNodeState.UNKNOWN_STATE),
+        SlurmNode(name="node-115", partition="main", state=SlurmNodeState.UNKNOWN_STATE),
```
Comment (TaekyungHeo) [inline tests/test_slurm_system.py:11]: I see you made an assumption about node names. You assume that node names follow the format node-[num]. However, this is not always guaranteed. We need to cover cases with node[num] as well. I am curious if the updated code covers this case too.

File: tests/test_slurm_system.py [hunk 3]
```diff
@@ -48,6 +54,20 @@ def test_parse_sinfo_output(slurm_system):
     slurm_system.parse_sinfo_output(sinfo_output, node_user_map)
     assert slurm_system.partitions["main"][0].state == SlurmNodeState.IDLE
     assert slurm_system.partitions["main"][1].state == SlurmNodeState.IDLE
+    
+def test_parse_sinfo_output2(slurm_system):
```
Comment (TaekyungHeo) [inline tests/test_slurm_system.py:58]: Let's use more descriptive function names for test cases as well. Names like test1, test2, and test3 are not good choices.

**Review by TaekyungHeo — CHANGES_REQUESTED** [review]

**Comment by TaekyungHeo** [comment]
Please fix failing CI pipelines as well.

File: tests/test_slurm_system.py [hunk 4]
```diff
@@ -37,26 +43,51 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
-    node_user_map = {"nodeA001": "root", "nodeB001": "user"}
+    sinfo_output = """
+    PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
+    main    up    3:00:00      1  inval node-081
+    main    up    3:00:00      5  drain node-[065-066,114,124-125]
+    main    up    3:00:00      2   resv node-[034-035]
+    main    up    3:00:00     88  alloc node-[033,036-064,067-080,082-113,115-123,126-128]
+    backup    up   12:00:00     16  alloc node-[01-16]
+    """
+    node_user_map = {'': 'user1',
```
Comment (amaslenn) [inline tests/test_slurm_system.py:55]: Please run `ruff format .` locally or set up your editor to do this on save.

File: tests/test_slurm_system.py [hunk 5]
```diff
@@ -37,26 +43,51 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
+    assert slurm_system.partitions["main"][1].state == SlurmNodeState.ALLOCATED
 
 
 @patch("cloudai.schema.system.SlurmSystem.get_squeue")
 @patch("cloudai.schema.system.SlurmSystem.get_sinfo")
 def test_update_node_states_with_mocked_outputs(mock_get_sinfo, mock_get_squeue, slurm_system):
-    mock_get_squeue.return_value = "nodeA001|root"
-    mock_get_sinfo.return_value = "PARTITION AVAIL TIMELIMIT NODES STATE NODELIST\n" "main up infinite 1 idle nodeA001"
+    mock_get_squeue.return_value = "node-115|user1"
+    mock_get_sinfo.return_value = "PARTITION AVAIL TIMELIMIT NODES STATE NODELIST\n" "main up infinite 1 idle node-115"
```
Comment (amaslenn) [inline tests/test_slurm_system.py:88]: IMO we should keep both.

```py

@pytest.mark.parametrize("sinfo,squeue,expecter_user", [
    ("PARTITION AVAIL TIMELIMIT NODES STATE NODELIST\n" "main up infinite 1 idle nodeA001", "nodeA001|root", "root"),
    ("PARTITION AVAIL TIMELIMIT NODES STATE NODELIST\n" "main up infinite 1 idle node-115", "node-115|user1", "user1"),
])
def test_funtion(sinfo: str, squeue: str, expected_user: str) -> None:
    ...
```

**Review by amaslenn — CHANGES_REQUESTED** [review]

File: tests/test_slurm_system.py (hunk 1 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py]: yes this is the output we get from IL1.
If you zant to change the tests you can do so but please approve my MR first

File: tests/test_slurm_system.py (hunk 2 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:11]: the code I put is just the older version of the function. What's important is that it works for IL1 (while your code doesnt)
Once again, you can approve the MR and then change the function as you please but check that it still works for IL1

Comment (lappazos) [reply tests/test_slurm_system.py:11]: @jeffnvidia can't you just add this other case here as well? does it add any more complexity?

File: tests/test_slurm_system.py (hunk 3 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py]: the old one was deleted now

File: tests/test_slurm_system.py (hunk 4 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py]: done

File: tests/test_slurm_system.py (hunk 2 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:11]: I covered both cases now

File: tests/test_slurm_system.py (hunk 5 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:94]: both are covered now

File: tests/test_slurm_system.py [hunk 6]
```diff
@@ -37,26 +43,51 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
+                    'node-11': 'user5', 
+                    'node-12': 'user5', 
+                    'node-13': 'user5', 
+                    'node-14': 'user5', 
+                    'node-15': 'user5', 
+                    'node-16': 'user5'}
     slurm_system.parse_sinfo_output(sinfo_output, node_user_map)
-    assert slurm_system.partitions["main"][0].state == SlurmNodeState.IDLE
-    assert slurm_system.partitions["main"][1].state == SlurmNodeState.IDLE
+    assert slurm_system.partitions["main"][0].state == SlurmNodeState.ALLOCATED
```
[the remark this answers is not in the document]
Comment (jeffnvidia) [reply tests/test_slurm_system.py]: done

File: src/cloudai/schema/system/slurm/slurm_system.py [hunk 7]
```diff
@@ -49,7 +49,7 @@ class SlurmSystem(System):
     """
 
     @classmethod
-    def parse_node_list(cls, node_list: List[str]) -> List[str]:
+    def parse_node_list(cls, node_list: str) -> List[str]:
```
Comment (amaslenn) [inline src/cloudai/schema/system/slurm/slurm_system.py:52]: Please add unit tests just for this function:
```py
@pytest.mark.parametrize(...)
def test_parse_node_list(input: str, expected: List[str]):
    ...
```

So far I find its code very confusing and not following the documentation of the function.

File: src/cloudai/schema/system/slurm/slurm_system.py [hunk 8]
```diff
@@ -589,20 +580,22 @@ def parse_sinfo_output(self, sinfo_output: str, node_user_map: Dict[str, str]) -
...
-                        for node in nodes:
-                            if node.name == node_name:
-                                node.state = state_enum
-                                node.user = node_user_map.get(node_name, "N/A")
+            node_names = self.parse_node_list(nodelist)
+
+            # Convert state to enum, handling states with suffixes
+            state_enum = self.convert_state_to_enum(state)
+
+            nodelist.split(",")
```
Comment (amaslenn) [inline src/cloudai/schema/system/slurm/slurm_system.py:588]: What is this line for?

File: src/cloudai/parser/system_parser/slurm_system_parser.py [hunk 9]
```diff
@@ -89,7 +89,9 @@ def safe_int(value):
                 raise ValueError("Partition data does not include a 'name' field.")
 
             raw_nodes = partition_data.get("nodes", [])
-            node_names = set(SlurmSystem.parse_node_list(raw_nodes))
```
Comment (amaslenn) [inline src/cloudai/parser/system_parser/slurm_system_parser.py:92]: Why not to parse all groups inside `parse_node_list`? It would be easier to use this function. Internally, you can have `parse_one_group` if that is useful.

File: src/cloudai/schema/system/slurm/slurm_system.py [hunk 10]
```diff
@@ -567,10 +560,8 @@ def parse_squeue_output(self, squeue_output: str) -> Dict[str, str]:
 
                 node_list_part, user = parts[0], "|".join(parts[1:])
                 # Handle cases where multiple node groups or ranges are specified
-                node_groups = node_list_part.split(",")
-                for node_group in node_groups:
-                    # Process each node or range using parse_node_list
-                    for node in self.parse_node_list([node_group.strip()]):
+                if node_list_part:
+                    for node in self.parse_node_list(node_list_part.strip()):
```
Comment (amaslenn) [inline src/cloudai/schema/system/slurm/slurm_system.py:564]: IMO `parse_node_list` should correctly handle empty output and return an empty list.

**Review by amaslenn — CHANGES_REQUESTED** [review]

File: src/cloudai/schema/system/slurm/slurm_system.py (hunk 7 above)
Comment (jeffnvidia) [reply src/cloudai/schema/system/slurm/slurm_system.py:52]: how is it confusing ? It's just parsing the outputs of sinfo.
Anyways, I just fixed the function, the description and its use are the same as before.

I'll add a unittest

Comment (amaslenn) [reply src/cloudai/schema/system/slurm/slurm_system.py:52]: Function arg is defined as a string: `node_list: str`.
But documentation says it should be a list of strings. So I wonder how is that even possible that input is different, but usage wasn't updated. I hope unit-tests will clarify it.

Comment (jeffnvidia) [reply src/cloudai/schema/system/slurm/slurm_system.py:52]: ah yes this part is my mistake it used to be list[str] now its str, I changed the arg but forgot to change the description
I'm updating now

File: src/cloudai/schema/system/slurm/slurm_system.py (hunk 8 above)
Comment (jeffnvidia) [reply src/cloudai/schema/system/slurm/slurm_system.py:588]: right this was in the old version, forgot to delete it

File: src/cloudai/parser/system_parser/slurm_system_parser.py (hunk 9 above)
Comment (jeffnvidia) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: The function parse_node_list is used in a lot of places and a lot of time, it's not used with a loop.

It is used with a loop at just 2 places so I think it's fine to keep this loop outside the function.

@TaekyungHeo what do you think ?

File: tests/test_slurm_system.py [hunk 11]
```diff
@@ -39,24 +55,73 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
+    )
 
     slurm_system.update_node_states()
+    for node in slurm_system.partitions["backup"]:
+        if node.name == "node01":
+            assert node.state == SlurmNodeState.ALLOCATED
+            assert node.user == "root"
+
+
+def test_parse_node_list(slurm_system):
```
Comment (amaslenn) [inline tests/test_slurm_system.py:114]: Please use `@pytest.mark.parametrize`:
```py
@pytest.mark.parametrize("input,expected", [
    ("node-[048-051]", ["node-048", "node-049", "node-050", "node-051"]),
    ...
])
def test_parse_node_list(input: str, expected: List[str]):
    parsed_nodes = slurm_system.parse_node_list(input)
    assert parsed_nodes == expected
```

File: src/cloudai/schema/system/slurm/slurm_system.py [hunk 12]
```diff
@@ -49,41 +49,35 @@ class SlurmSystem(System):
...
-                    if "-" in r:
-                        expanded_nodes.extend([f"{prefix}{str(i).zfill(max_length)}" for i in range(start, end + 1)])
-                    else:
-                        # For single nodes, append directly with appropriate padding.
-                        expanded_nodes.append(f"{prefix}{start_str.zfill(max_length)}")
-            else:
-                expanded_nodes.append(node)
-        return expanded_nodes
+        nodes = []
+        if node_list:
```
Comment (amaslenn) [inline src/cloudai/schema/system/slurm/slurm_system.py:64]: To reduce nesting, you can handle this corner case explicitly:
```py
if not node_list:
    return []

...
```

File: tests/test_slurm_system.py [hunk 13]
```diff
@@ -39,24 +55,73 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
+    node_lists = ["node-[048-051]", "node-[055,114]", "", "node-001", "node[1-4]"]
+
+    expected_outputs = [
```
Comment (amaslenn) [inline tests/test_slurm_system.py:117]: Worth adding: `node-name[01-03,05-08,10]`.

File: tests/test_slurm_system.py [hunk 14]
```diff
@@ -39,24 +55,73 @@ def test_parse_squeue_output_with_node_ranges_and_root_user(slurm_system):
...
+        "node05": "user5",
+        "node06": "user5",
+        "node07": "user5",
+        "node08": "user5",
+    }
     slurm_system.parse_sinfo_output(sinfo_output, node_user_map)
-    assert slurm_system.partitions["main"][0].state == SlurmNodeState.IDLE
-    assert slurm_system.partitions["main"][1].state == SlurmNodeState.IDLE
+    for i in range(len(slurm_system.partitions["main"])):
+        assert slurm_system.partitions["main"][i].state == SlurmNodeState.ALLOCATED
```
Comment (amaslenn) [inline tests/test_slurm_system.py:85]: According to your `sinfo_output`, nodes have different states, but after parsing all of them are allocated. Is that expected?

File: src/cloudai/schema/system/slurm/slurm_system.py [hunk 15]
```diff
@@ -567,11 +561,8 @@ def parse_squeue_output(self, squeue_output: str) -> Dict[str, str]:
 
                 node_list_part, user = parts[0], "|".join(parts[1:])
                 # Handle cases where multiple node groups or ranges are specified
-                node_groups = node_list_part.split(",")
-                for node_group in node_groups:
-                    # Process each node or range using parse_node_list
-                    for node in self.parse_node_list([node_group.strip()]):
-                        node_user_map[node] = user.strip()
+                for node in self.parse_node_list(node_list_part.strip()):
```
Comment (amaslenn) [inline src/cloudai/schema/system/slurm/slurm_system.py:564]: You can call `.strip()` during initialization to avoid repeating it in the loop.

**Review by amaslenn — CHANGES_REQUESTED** [review]

File: src/cloudai/schema/system/slurm/slurm_system.py (hunk 12 above)
Comment (jeffnvidia) [reply src/cloudai/schema/system/slurm/slurm_system.py:64]: sure I updated

File: tests/test_slurm_system.py (hunk 14 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:85]: according to sinfo_output, all nodes in main are allocated and all nodes in backup are idle so the behaviour seems good for me

File: src/cloudai/schema/system/slurm/slurm_system.py (hunk 15 above)
Comment (jeffnvidia) [reply src/cloudai/schema/system/slurm/slurm_system.py:564]: updated

File: tests/test_slurm_system.py (hunk 13 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:117]: added

File: tests/test_slurm_system.py (hunk 11 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py:114]: I updated but then should all the tests look like this ?
Because the rest was done in another way

File: tests/test_slurm_system.py (hunk 14 above)
Comment (amaslenn) [reply tests/test_slurm_system.py:85]: Are we talking about the same code? :)

```py
    sinfo_output = """PARTITION AVAIL  TIMELIMIT  NODES  STATE NODELIST
    main    up    3:00:00      1  inval node-081                      # <- NOT ALLOCATED
    main    up    3:00:00      5  drain node-[065-066,114,124-125]    # <- NOT ALLOCATED
    main    up    3:00:00      2   resv node-[034-035]                # <- NOT ALLOCATED
    main    up    3:00:00     88  alloc node-[033,036-064,067-080,082-113,115-123,126-128]
    backup    up   12:00:00     16  idle node[01-16]
    """
```

Comment (jeffnvidia) [reply tests/test_slurm_system.py:85]: ok I see the misunderstanding
Indeed not all nodes are allocated in the example but if you look at the slurm system, we're just checking nodes 115 to 122 (in main) which are indeed allocated

File: src/cloudai/parser/system_parser/slurm_system_parser.py (hunk 9 above)
Comment (TaekyungHeo) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: @amaslenn , could you please review our comments and reply?

Comment (amaslenn) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: Current proposal is:
```py
node_names = set()
for group in raw_nodes:
    node_names.update(set(SlurmSystem.parse_node_list(group)))
```

I suggest changing it to:
```py
node_names = set(SlurmSystem.parse_node_list(raw_nodes))
```
And for this we need to move this loop into the `parse_node_list`.

I just don't understand why it is not suitable here. @jeffnvidia could you please explain?

File: tests/test_slurm_system.py (hunk 11 above)
Comment (amaslenn) [reply tests/test_slurm_system.py]: In other tests it is different, there is no loop by input-expected pairs.

File: src/cloudai/parser/system_parser/slurm_system_parser.py (hunk 9 above)
Comment (TaekyungHeo) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: cc: @lappazos

Comment (jeffnvidia) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: > Current proposal is:
> 
> ```python
> node_names = set()
> for group in raw_nodes:
>     node_names.update(set(SlurmSystem.parse_node_list(group)))
> ```
> 
> I suggest changing it to:
> 
> ```python
> node_names = set(SlurmSystem.parse_node_list(raw_nodes))
> ```
> 
> And for this we need to move this loop into the `parse_node_list`.
> 
> I just don't understand why it is not suitable here. @jeffnvidia could you please explain?

Yes it is possible of courser.
Once again, I'm just saying that in 3 other places we use this function without a list so I would need to create an artificial list with one element, I think it wouldnt look good.
@amaslenn

Comment (amaslenn) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: Do you mean this:
```py
def parse_node_list(nodes: Union[List[str], str], ...):
    all_nodes = nodes
    if isinstance(nodes, str):
        all_nodes = [nodes]
    ...
```
?

Here is a more testable solution:
```py
def parse_one_node_list(nodes: str, ...) -> ...:
    ...

def parse_node_list(nodes: Union[List[str], str], ...) -> ...:
    all_nodes: List[str] = []
    if isinstance(nodes, list):
        all_nodes.extend(self.parse_one_node_list(...))
    else:
        all_nodes = self.parse_one_node_list(...)
    return all_nodes
```

Comment (jeffnvidia) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: are you sure this is really needed for just 2 loops ?
It looks overly complicated in my opinion

File: tests/test_slurm_system.py (hunk 14 above)
Comment (jeffnvidia) [reply tests/test_slurm_system.py]: updated

File: src/cloudai/parser/system_parser/slurm_system_parser.py (hunk 9 above)
Comment (amaslenn) [reply src/cloudai/parser/system_parser/slurm_system_parser.py:92]: OK, let's keep external loops.

**Review by amaslenn — APPROVED** [review]

**Review by TaekyungHeo — APPROVED** [review]
